File: Python/example_30/pred_prey.py
import numpy as np
from scipy.integrate import odeint


############################
### Case study 30
############################
def pred_prey(x, plugin):
    """
    This function runs the 1-predator-1-prey model.
    """
    # Solve the system of differential equations using odeint
    u = odeint(PP_ode_equations, plugin['u0'], plugin['t'], args=(x,))
    # Two columns of predator and prey abundances
    # Need to re-order the nx2 matrix to yield two columns underneath eachother
    u = u.T 
    # Return the simulated abundances of both species as one vector
    return u.flatten()


def PP_ode_equations(u, t, pars):
    """
    Solves the coupled ordinary differential equations of the 1-predator-1-prey population model.
    """
    # Initial states
    x = u[0]            # Prey population
    y = u[1]            # Predator population
    
    # Parameters
    r = pars[0]         # Intrinsic rate of prey natural increase
    alfa = pars[1]      # Proportionality constant linking prey mortality to prey and predator numbers
    m = pars[2]         # Mortality rate of predators
    theta = pars[3]     # Proportionality constant linking predator increase to prey and predator numbers
    K = 50              # Maximum number of preys (carrying capacity)
    
    # Ordinary differential equations
    dxdt = r * x * (1 - x / K) - alfa * x * y
    dydt = -m * y + theta * x * y
    
    # Return the derivatives as a vector
    return [dxdt, dydt]

File: MATLAB/example_30/pred_prey.m
function Y = pred_prey(x,plugin)
% This function runs the 1-predator-1-prey model

% use built-in ode45 function to solve 2 differential equations from t = 0 to t = 335
[ ~ , u ] = ode45(@(t,u) PP(t,u,x),plugin.t,plugin.u0,plugin.ode_options);
% now return the simulated abundances of both species as one vector
Y = u(:);
end

% Secundairy function
function u_out = PP(t,u,pars)
% Solves the coupled ordinary differential equations of the 1-predator-1-prey population model

% Assign the states
x = u(1); y = u(2);

% intrinsic rate of prey natural increase
r = pars(1);
% proportionality constant linking prey mortality to the number of prey and predators
alfa = pars(2);
% mortality rate of predators
m = pars(3);
%  proportionality constant linking the increase in predators to the number of predators and prey
theta = pars(4);
% maximum number of preys that the environment can support (carrying capacity?)
K = 50;

% Now compute each of the two ordinary differential equations
dxdt = r * x * ( 1 - x/K ) - alfa * x * y;
dydt = -m * y + theta * x * y;

% Now assemble the dx/dt and dy/dt in one vector u
u_out = [ dxdt ; dydt ];
end